### ISSUE TITLE
IndentationCheck: Wrong indention in chained methods

### ISSUE THREAD

**Issue by karolhor** [checkstyle/checkstyle#4249]
Hi, I have problems with indention level in chained methods.

/var/tmp $ javac CheckstyleExample.java

/var/tmp $ cat CheckstyleExample.java

```java
package example;

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString()
        ); // this is reported as incorrect despite alignment to first chain
  }
}
```

/var/tmp $ cat config.xml

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <module name="TreeWalker">
        <module name="Indentation">
            <property name="basicOffset" value="2"/>
            <property name="braceAdjustment" value="0"/>
            <property name="caseIndent" value="2"/>
            <property name="throwsIndent" value="4"/>
            <property name="lineWrappingIndentation" value="4"/>
            <property name="arrayInitIndent" value="2"/>
        </module>
    </module>
</module>
```

/var/tmp $ java -jar checkstyle-7.6.1-all.jar -c config.xml CheckstyleExample.java

```bash
Starting audit...
[ERROR] /path/to/example/CheckstyleExample.java:12: 'method call rparen' have incorrect indentation level 8, expected level should be 6. [Indentation]
Audit done.
Checkstyle ends with 1 errors.
```

---------------

As I've mentioned in a comment in example, the align indention of right `)` is reported as incorrect, but this examples are valid:

```java
package com.pearson.adminpanel;

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString());  // no violation, but this is not what I would like to achieve
  }
}
```
```java
package com.pearson.adminpanel;

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString()
      );  // no violation, but looks strange in a middle of above indentions
  }
}
```
```java
package com.pearson.adminpanel;

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString()
        ) // no violation, but only in a middle of chaining
        .append("bbb"); 
  }
}
```

I think there could be an extra rule for fixing this problem.

**Comment by romani** [comment]
> // correct

I am not sure what does this mean .... probably "no violation" . Please fix to make it clear.

**Comment by karolhor** [comment]
@romani, I've changed description as you suggested: from "correct" to "no violation"

**Comment by ps-sp** [comment]
I am on it.

**Comment by leventov** [comment]
Any updates here?

**Comment by romani** [comment]
I removed comment of "ps-sp", he is not doing any progress here.
Unfortunately there is no progress in issues.
If you affected by it and have some time to resolve it , please be welcome to contribute, we could guide on development process.

**Comment by aecepoglu** [comment]
Seems this issue is still open. I would be willing to resolve it if nobody else is on it.

**Comment by rnveach** [comment]
@aecepoglu You are free to work on this issue, but be warned indentation is one of our most complex checks.

**Comment by Harshit-Dandriyal** [comment]
I think I should give it a try.

**Comment by Harshit-Dandriyal** [comment]
How can I start approaching this problem?

**Comment by Scrayos** [comment]
I additionally found out that this is only reported if the statement is not part of an assignment. If anyone would start working on this, maybe this information could be of help.
```java
    public void test() {
        // no violation – correct
        Optional<Aura<JCBungeeUser>> aura = JCBungee.getInstance().getAuraHandler()
            .getImplementation(
                AuraType.Z62
            );
        
        // violation – should not be one
        JCBungee.getInstance().getAuraHandler()
            .getImplementation(
                AuraType.Z62
            );
    }
```

**Comment by aecepoglu** [comment]
I had some development planned for this issue, but it has been too long now and I will not be attempting to solve this.

**Comment by Abhishek-kumar09** [comment]
@Scrayos, do you remember the config for the code to check against.
It would be better if we could test your code too for resolution.

About @karolhor tests this issue is seemed to be **fixed**:
```java
package example;

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString()
        ); // NO VIOLATION
  }
}

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString());  // no violation, but this is not what I would like to achieve
  }
}

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString()
      );  // THERE IS THE VIOLATION NOW, BECAUSE THE PAREN IS NOT ALIGNED line:37
  }
}

public class CheckstyleExample {
  public void execute() {
    StringBuilder sb = new StringBuilder();

    sb.append("aaa")
        .append(new StringBuilder()
            .append("1")
            .append("2")
            .toString()
        ) // no violation, but only in a middle of chaining
        .append("bbb"); 
  }
}
```
```
<module name="Checker">
  <property name="charset" value="UTF-8" />
  <module name="TreeWalker">
    <module name="Indentation">
         <property name="basicOffset" value="2"/>
            <property name="braceAdjustment" value="0"/>
            <property name="caseIndent" value="2"/>
            <property name="throwsIndent" value="4"/>
            <property name="lineWrappingIndentation" value="4"/>
            <property name="arrayInitIndent" value="2"/>
    </module>
  </module>
</module>
```
output:
```
java -jar checkstyle-8.36.2-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /home/invisible-a/playGround/java/Test.java:37:7: 'method call rparen' has incorrect indentation level 6, expected level should be 8. [Indentation]
```

**Comment by romani** [comment]
I close issue, as it is not reproduced on latest release. Some fix fixed issue
